File: src/main/java/io/github/mg138/tsbook/utils/Translate.java
package io.github.mg138.tsbook.utils;

import io.github.mg138.tsbook.util.RGBUtil;
import me.clip.placeholderapi.PlaceholderAPI;
import org.bukkit.OfflinePlayer;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.file.YamlConfiguration;

import java.util.*;

public class Translate {
    private final YamlConfiguration language;

    public Translate(YamlConfiguration lang) {
        language = lang;
    }

    public String translateString(String string) {
        return translateString(string, null);
    }

    public String translateString(String string, OfflinePlayer player) {
        while (PlaceholderAPI.containsPlaceholders(string)) {
            String newString = PlaceholderAPI.setPlaceholders(player, string);
            if (string.equals(newString)) throw new IllegalArgumentException("Placeholder stuck in a loop.");
            string = newString;
        }
        return RGBUtil.INSTANCE.translate(string);
    }

    public String translate(String path) {
        return translate(path, null, language);
    }

    public String translate(String path, OfflinePlayer player) {
        return translate(path, player, language);
    }

    public String translate(String path, OfflinePlayer player, YamlConfiguration file) {
        return translateString(file.getString(path), player);
    }

    public String translate(String path, OfflinePlayer player, ConfigurationSection section) {
        return translateString(section.getString(path), player);
    }

    public List<String> translateList(String path) {
        return translateList(path, null, language);
    }

    public List<String> translateList(String path, OfflinePlayer player) {
        return translateList(path, player, language);
    }

    public List<String> translateList(String path, OfflinePlayer player, YamlConfiguration file) {
        return translateList(path, player, Objects.requireNonNull(file.getConfigurationSection("")));
    }

    public List<String> translateList(String path, OfflinePlayer player, ConfigurationSection section) {
        List<String> list = section.getStringList(path);

        if (list.isEmpty()) {
            try {
                String string = section.getString(path);
                if (string == null) return null;
                return translateByList(Arrays.asList(string.split("\\\\n|\\n")), player);
            } catch (NullPointerException e) {
                e.printStackTrace();
                return null;
            }
        }
        return translateByList(list, null);
    }

    public List<String> translateByList(List<String> list) {
        return translateByList(list, null);
    }

    public List<String> translateByList(List<String> list, OfflinePlayer player) {
        List<String> result = new ArrayList<>();
        try {
            for (String string : list) {
                result.addAll(Arrays.asList(translateString(string, player).split("\\\\n|\\n")));
            }
        } catch (NullPointerException e) {
            e.printStackTrace();
            return Collections.emptyList();
        }
        return result;
    }
}

File: src/main/java/io/github/mg138/tsbook/utils/Translate.kt
package io.github.mg138.tsbook.utils

import io.github.mg138.tsbook.util.RGBUtil
import me.clip.placeholderapi.PlaceholderAPI
import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.configuration.file.YamlConfiguration
import java.util.*

class Translate(private val language: YamlConfiguration) {
    fun translateString(string: String, player: OfflinePlayer? = null): String {
        var copy = string
        var newString: String
        while (PlaceholderAPI.containsPlaceholders(copy)) {
            newString = PlaceholderAPI.setPlaceholders(player, copy)
            require(copy != newString) {
                """Placeholder stuck in a loop. (Does it contain placeholders that don't exist?)
		           The error string: $newString"""
            }
            copy = newString
        }
        return RGBUtil.translate(copy)
    }

    fun translate(path: String, player: OfflinePlayer? = null, file: YamlConfiguration = language): String {
        return translateString(file.getString(path)!!, player)
    }

    fun translate(path: String, player: OfflinePlayer?, section: ConfigurationSection): String {
        return translateString(section.getString(path)!!, player)
    }

    fun translateList(path: String, player: OfflinePlayer? = null, file: YamlConfiguration = language): List<String> {
        return translateList(path, player, file.getConfigurationSection("")!!)
    }

    fun translateList(path: String, player: OfflinePlayer?, section: ConfigurationSection): List<String> {
        val list = section.getStringList(path)
        return when {
            list.isEmpty() -> {
                try {
                    val string = section.getString(path) ?: return emptyList()
                    translateByList(string.split("\\\\n|\\n"), player)
                } catch (e: NullPointerException) {
                    e.printStackTrace()
                    return emptyList()
                }
            }
            else -> translateByList(list, player)
        }
    }

    fun translateByList(list: List<String>, player: OfflinePlayer? = null): List<String> {
        val result: MutableList<String> = ArrayList()
        try {
            list.forEach { string ->
                result.addAll(translateString(string, player).split("\\\\n|\\n"))
            }
        } catch (e: NullPointerException) {
            e.printStackTrace()
            return emptyList()
        }
        return result
    }
}